pkg/search/postgres/storecache.go

```diff
@@ -0,0 +1,329 @@
+// subStore implements subset of Store operations.
+func NewGenericStoreWithCache[T any, PT unmarshaler[T]](
+	db postgres.DB,
+	schema *walker.Schema,
+	pkGetter primaryKeyGetter[T, PT],
+	insertInto inserter[T, PT],
+	copyFromObj copier[T, PT],
+	setAcquireDBConnDuration durationTimeSetter,
+	setPostgresOperationDurationTime durationTimeSetter,
+	upsertAllowed upsertChecker[T, PT],
+	targetResource permissions.ResourceMetadata,
+) Store[T, PT] {
+	underlyingStore := NewGenericStore[T, PT](
+		db,
+		schema,
+		pkGetter,
+		insertInto,
+		copyFromObj,
+		setAcquireDBConnDuration,
+		setPostgresOperationDurationTime,
+		upsertAllowed,
+		targetResource,
+	)
+	store := &CachedStore[T, PT]{
+		schema:          schema,
+		pkGetter:        pkGetter,
+		targetResource:  targetResource,
+		cache:           make(map[string]PT),
+		useCache:        true,
+		underlyingStore: underlyingStore,
+	}
+	store.cacheLock.Lock()
+	defer store.cacheLock.Unlock()
+	store.resetCacheNoLock()
+	return store
+}
+
+// NewGenericStoreWithPermissionCheckerWithCache returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStoreWithPermissionCheckerWithCache[T any, PT unmarshaler[T]](
+	db postgres.DB,
+	schema *walker.Schema,
+	pkGetter primaryKeyGetter[T, PT],
+	insertInto inserter[T, PT],
+	copyFromObj copier[T, PT],
+	setAcquireDBConnDuration durationTimeSetter,
+	setPostgresOperationDurationTime durationTimeSetter,
+	checker walker.PermissionChecker,
+) Store[T, PT] {
+	underlyingStore := NewGenericStoreWithPermissionChecker[T, PT](
+		db,
+		schema,
+		pkGetter,
+		insertInto,
+		copyFromObj,
+		setAcquireDBConnDuration,
+		setPostgresOperationDurationTime,
+		checker,
+	)
+	store := &CachedStore[T, PT]{
+		schema:            schema,
+		pkGetter:          pkGetter,
+		permissionChecker: checker,
+		cache:             make(map[string]PT),
+		useCache:          true,
+		underlyingStore:   underlyingStore,
+	}
+	store.cacheLock.Lock()
+	defer store.cacheLock.Unlock()
+	store.resetCacheNoLock()
+	return store
+}
+
+// CachedStore implements subset of Store interface for resources with single ID.
+type CachedStore[T any, PT unmarshaler[T]] struct {
+	mutex                         sync.RWMutex
+	schema                        *walker.Schema
+	pkGetter                      primaryKeyGetter[T, PT]
+	setCacheOperationDurationTime durationTimeSetter
+	permissionChecker             walker.PermissionChecker
+	targetResource                permissions.ResourceMetadata
+	underlyingStore               Store[T, PT]
+	cache                         map[string]PT
+	useCache                      bool
+	cacheLock                     sync.RWMutex
+}
+
+// Upsert saves the current state of an object in storage.
+func (c *CachedStore[T, PT]) Upsert(ctx context.Context, obj PT) error {
+	dbErr := c.underlyingStore.Upsert(ctx, obj)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	if dbErr != nil {
```

Comment:
Why do we need to reset the cache when the is a db error? These are transactions so it should be everything or nothing?